pkg/resource/deploy/step_generator.go

```diff
@@ -34,13 +34,65 @@ type stepGenerator struct {
 	opts Options // options for this step generator
 
 	urns     map[resource.URN]bool // set of URNs discovered for this plan
+	reads    map[resource.URN]bool // set of URNs read for this plan
 	deletes  map[resource.URN]bool // set of URNs deleted in this plan
 	replaces map[resource.URN]bool // set of URNs replaced in this plan
 	updates  map[resource.URN]bool // set of URNs updated in this plan
 	creates  map[resource.URN]bool // set of URNs created in this plan
 	sames    map[resource.URN]bool // set of URNs that were not changed in this plan
 }
 
+// GenerateReadSteps is responsible for producing one or more steps required to service
+// a ReadResourceEvent coming from the language host.
+func (sg *stepGenerator) GenerateReadSteps(event ReadResourceEvent) ([]Step, error) {
+	urn := sg.generateURN(event.Parent(), event.Type(), event.Name())
+	newState := resource.NewState(event.Type(),
+		urn,
+		true,  /*custom*/
+		false, /*delete*/
+		event.ID(),
+		event.Properties(),
+		make(resource.PropertyMap), /* outputs */
+		event.Parent(),
+		false, /*protect*/
+		true,  /*external*/
+		event.Dependencies(),
+		nil /* initErrors */)
+	old, hasOld := sg.plan.Olds()[urn]
+
+	// If the snapshot has an old resource for this URN and it's not external, we're going
+	// to have to delete the old resource and conceptually replace it with the resource we
+	// are about to read.
+	//
+	// We accomplish this through the "read-replacement" step, which atomically reads a resource
+	// and marks the resource it is replacing as pending deletion.
+	//
+	// In the event that the new "read" resource's ID matches the existing resource,
```

Comment:
This surprised me. So, are you saying we won't physically delete the resource, when you change your program from `new`ing a resource to `get`ting it, provided the IDs match identically? This makes me a little nervous, as it's not just composition of existing operations.

Reply:
Yes - this is a new case, and something @pgavlin @swgillespie and I spoke about in person at length this week.  

There are two things that convince me this design makes sense:
1. It supports a really valuable scenario - detach management of an existing resource from Pulumi without destroying the resource.  For example, this would be useful when splitting one Pulumi stack into multiple stacks and factoring some resources out of the original stack to be `.gets`.
2. If we didn't do this special case behaviour, the operation would be guaranteed to fail - we would delete the resource immediately before (or worse, immediately after - or even worse, non-deterministically before/after) the `.get` tries to capture a handle to it.  It's hard to imagine what benefit it provides to allow this to happen, especially given that (1) is a more useful interpretation.